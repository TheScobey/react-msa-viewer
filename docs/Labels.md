# `Labels` (component)

Displays the sequence names.

## Props

### `engine`

Rendering engine: `canvas` or `webgl` (experimental).

type: `enum('canvas'|'webgl')`
defaultValue: `"canvas"`


### `font`

Font of the sequence labels, e.g. `20px Arial`

type: `string`


### `height`

Width of the component (in pixels), e.g. `100`

type: `number`
defaultValue: `100`


### `msecsPerFps`

defaultValue: `60`


### `msecsPerSecs`

Maximum number of frames per second, e.g. `1000 / 60`

type: `number`


### `style`

Custom style configuration.

type: `object`


### `width`

Width of the component (in pixels), e.g. `100`

type: `number`
defaultValue: `100`

