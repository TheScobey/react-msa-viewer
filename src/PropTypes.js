/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import PropTypes from 'prop-types';

/**
 * Definition of a single sequence object.
 *   name: label or id of the sequence (doesn't need to be unique)
 *   sequence: raw sequence data (e.g. AGAAAA)
 */
export const SequencePropType = PropTypes.shape({
  name: PropTypes.string,
  sequence: PropTypes.string,
})

export const AllowedColorschemes = [
  "buried_index", "clustal", "clustal2", "cinema",
  "helix_propensity", "hydro",
  "lesk", "mae", "nucleotide", "purine_pyrimidine",
  "strand_propensity", "taylor", "turn_propensity", "zappo",
];

export const ColorSchemePropType = PropTypes.oneOf(AllowedColorschemes);

export const PositionPropType = PropTypes.shape({
  xPos: PropTypes.number,
  yPos: PropTypes.number,
})

export const MSAPropTypes = {
  /**
   * Sequence data.
   */
  sequences: PropTypes.arrayOf(SequencePropType).isRequired,

  /**
   * Width of the sequence viewer (in pixels), e.g. `500`.
   */
  width: PropTypes.number,

  /**
   * Height of the sequence viewer (in pixels), e.g. `500`.
   */
  height: PropTypes.number,

  /**
   * Width of the main tiles (in pixels), e.g. `20`
   */
  tileWidth: PropTypes.number,

  /**
   * Height of the main tiles (in pixels), e.g. `20`
   */
  tileHeight: PropTypes.number,

  /**
   * Font of the individual residue tiles, e.g. `"20px Arial"`.
   */
  tileFont: PropTypes.string,

  /**
   * Maximum number of frames per second, e.g. `10000 / 60`.
   */
  msecsPerFps: PropTypes.number,

  /**
   * Current x and y position of the viewpoint
   * in the main sequence viewer (in pixels).
   * This specifies the position of the top-left corner
   * of the viewpoint within the entire alignment,
   * e.g. `{xPos: 20, yPos: 5}`.
   */
  position: PositionPropType,

  /**
   * Colorscheme to use. Currently the follow colorschemes are supported:
   * `buried_index`, `clustal`, `clustal2`, `cinema`, `helix_propensity`, `hydro`,
   *`lesk`, `mae`, `nucleotide`, `purine_pyrimidine`, `strand_propensity`, `taylor`,
   * `turn_propensity`, and `zappo`.
   *
  * See [msa-colorschemes](https://github.com/wilzbach/msa-colorschemes) for details.
  */
  colorScheme: ColorSchemePropType,

  /**
   * Background color to use, e.g. `red`
   */
  backgroundColor: PropTypes.string,

  /**
   * Rendering engine: `canvas` or `webgl` (experimental).
   */
  engine: PropTypes.oneOf(['canvas', 'webl']), // experimental
};


// TODO: separate individual properties into their components
export const msaDefaultProps = {
  width: 500,
  height: 100,
  tileWidth: 20,
  tileHeight: 20,
  tileFont: "20px Arial",
  msecsPerFps: 1000  / 60,
  position: {
    xPos: 0,
    yPos: 0,
  },
  colorScheme: "clustal",
  backgroundColor: "red",
  engine: "canvas", // experimental
};

export {PropTypes};
