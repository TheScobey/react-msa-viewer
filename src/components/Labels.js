/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import React, { Component } from 'react';
import { connect } from 'react-redux'

import { throttle } from 'lodash';

import propsToRedux from '../store/propsToRedux';

import Canvas from '../drawing/canvas';

class LabelsComponent extends Component {

  static defaultProps = {
    width: 80, // TODO: can we calculate this automatically?
    style: {display: "block"},
  }

  constructor(props) {
    super(props);
    this.canvas = React.createRef();
    this.draw = throttle(this.draw, this.props.viewpoint.msecsPerFps);
  }

  componentDidMount() {
    this.ctx = new Canvas(this.canvas.current);
    this.draw();
  }

  componentDidUpdate() {
    this.draw();
  }

  draw() {
    const [,tileHeight] = this.props.viewpoint.tileSizes;
    this.ctx.startDrawingFrame();
    let xPos = 0;
    let yPos = -this.props.position.yPos + 3;
    this.ctx.font(this.props.viewpoint.labelSize);
    for (let i = 0; i < this.props.nrSequences; i++) {
      let label;
      if (this.props.labels[i]) {
        label = this.props.labels[i];
      } else {
        label = "Sequence " + i;
      }
      this.ctx.fillText(label, xPos, yPos, this.props.width, tileHeight);
      yPos += tileHeight;
    }
    this.ctx.endDrawingFrame();
  }

  render() {
    return (
      <div style={this.props.style}>
        <canvas
          ref={this.canvas}
          width={this.props.width}
          height={this.props.viewpoint.height}
        >
        </canvas>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    position: state.position,
    viewpoint: state.viewpoint,
    nrSequences: state.sequences.raw.length,
    labels: state.sequences.raw.map(s => s.name),
  }
}

export default propsToRedux(connect(
  mapStateToProps,
)(LabelsComponent));
