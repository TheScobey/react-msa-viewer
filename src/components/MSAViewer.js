/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import React, { Component } from 'react';
import MSAProvider from '../store/provider';

import PositionBar from './PositionBar';
import SequenceViewer from './SequenceViewer';
import SequenceOverview from './SequenceOverview';
import OverviewBar from './OverviewBar';
import Labels from './Labels';

import createMSAStore from '../store/createMSAStore';
import propsToRedux from '../store/propsToRedux';

const labelsAndSequenceDiv = {
  display: "flex",
};

// TODO: support changing the store dynamically
// TODO: when props of children update -> update store
// TODO: support using the child components in stand-alone mode
class MSAViewerComponent extends Component {

  componentWillMount() {
    //this.store = this.props.store || createMSAStore(
      //pick(this.props, MSAViewerProps)
    //);
  }
  render() {
    const {children, msaStore, ...otherProps} = this.props;
    if (children) {
      console.log("children: ", children);
      return (
        <MSAProvider store={msaStore}>
          <div {...otherProps}>
            {children}
          </div>
        </MSAProvider>
      );
    } else {
      // TODO: add more advanced layouts
      const currentState = msaStore.getState();
      const labelsPadding = currentState.viewpoint.tileSizes[1];
      const overviewBarHeight = currentState.viewpoint.overviewBar.height;
      const labelsStyle = {
        paddingTop: labelsPadding + overviewBarHeight,
      }
      const separatorPadding = {
        height: 10,
      };
      return (
        <MSAProvider store={msaStore}>
          <div style={labelsAndSequenceDiv}>
            <Labels
              style={labelsStyle}
            />
            <div>
              <OverviewBar />
              <PositionBar />
              <SequenceViewer />
              <div style={separatorPadding} />
              <SequenceOverview />
            </div>
          </div>
        </MSAProvider>
      );
    }
  }
}

const MSAViewer = propsToRedux(MSAViewerComponent);

MSAViewer.PropTypes = {
  // store
  // sequences, required
};

// TODO: re-include propsToRedux here?
export default MSAViewer;
export {
  createMSAStore,
  Labels,
  MSAViewer,
  OverviewBar,
  PositionBar,
  SequenceOverview,
  SequenceViewer,
};
