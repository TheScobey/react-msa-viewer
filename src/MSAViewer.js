import React, { Component } from 'react';
import PropTypes from 'prop-types';

const schemes = new (require('msa-colorschemes'))();

class MSAViewer extends Component {

  constructor(props) {
    super(props);
    this.canvas = React.createRef();
  }

  componentDidMount() {
    this.ctx = this.canvas.current.getContext('2d');
    this.scheme = schemes.getScheme(this.props.scheme);
    this.drawSequences();
  }

  drawSequences() {
    const xSize = 20;
    const ySize = 20;
    let yPos = 0, xPos = 0;
    for (const seq of this.props.sequences) {
      for (const el of seq.sequence) {
        this.ctx.font = "20px Arial";
        this.ctx.fillStyle = this.scheme.getColor(el);
        this.ctx.fillRect(xPos, yPos, xSize, ySize);
        this.ctx.fillStyle = "#000000";
        this.ctx.fillText(el, xPos + 2, yPos + 17);
        xPos += xSize;
      }
      xPos = 0;
      yPos += ySize;
    }
  }

  componentDidUpdate() {

  }

  shouldComponentUpdate(newProps) {
    return this.props.target !== newProps.target;
  }

  render() {
    return (
      <canvas ref={this.canvas} with={500} height={500}>
      </canvas>
    );
  }
}

const SequencePropType = PropTypes.shape({
  name: PropTypes.string,
  sequence: PropTypes.string,
})

MSAViewer.propTypes = {
  color: PropTypes.string,
  scheme: PropTypes.string,
  sequences: PropTypes.arrayOf(SequencePropType).isRequired,
};

MSAViewer.defaultProps = {
  color: "red",
  scheme: "clustal",
}

export {MSAViewer};
export default MSAViewer;
