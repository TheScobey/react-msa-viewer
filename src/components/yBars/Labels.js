/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/
import React, { Component, PureComponent } from 'react';
import PropTypes from 'prop-types';
import shallowCompare from 'react-addons-shallow-compare';
import {
  pick
} from 'lodash-es';

import msaConnect from '../../store/connect'

import YBar from './yBar';

function createLabel({sequences, tileHeight, labelComponent,
  labelStyle, labelAttributes}) {
  /**
   * Displays an individual sequence name.
   */
  class Label extends PureComponent {
    render() {
      const {index, ...otherProps} = this.props;
      if (labelComponent) {
        const LabelComponent = labelComponent;
        return <LabelComponent sequence={sequences[index]} index={index} />;
      } else {
        otherProps.style = {
          ...this.props.style,
          height: tileHeight,
          ...labelStyle
        }
        const attributes = {...otherProps, ...labelAttributes};
        return (
          <div {...attributes} >
            {sequences[index].name}
          </div>
        );
      }
    }
  }
  return Label;
}

/**
 * Displays the sequence names.
 */
class HTMLLabelsComponent extends Component {

  constructor(props) {
    super(props);
    this.updateLabel();
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (["sequences", "tileHeight", "labelComponent"].some(key=> {
      return nextProps[key] !== this.props[key];
    }, true)){
      this.updateLabel();
      return true;
    }
    return shallowCompare(this, nextProps, nextState);
  }

  updateLabel() {
    this.cache = function(){};
    this.label = createLabel(pick(this.props, [
      "sequences", "tileHeight",
      "labelComponent", "labelStyle", "labelAttributes"
    ]));
  }

  render() {
    const {cacheElements,
      dispatch,
      labelComponent,
      labelStyle,
      labelAttributes,
      ...otherProps} = this.props;
    return (
      <YBar
        tileComponent={this.label}
        cacheElements={cacheElements}
        componentCache={this.cache}
        {...otherProps}
      />
    );
  }
}

HTMLLabelsComponent.defaultProps = {
  cacheElements: 10,
  labelStyle: {},
};

HTMLLabelsComponent.propTypes = {
  /**
   * Font of the sequence labels, e.g. `20px Arial`
   */
  font: PropTypes.string,

  /**
   * Component to create labels from.
   */
  labelComponent: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),

  /**
   * Inline styles to apply to the Label component
   */
  style: PropTypes.object,

  /**
   * Inline styles to apply to each label.
   */
  labelStyle: PropTypes.object,

  /**
   * Attributes to apply to each label.
   */
  labelAttributes: PropTypes.object,
}

const mapStateToProps = state => {
  return {
    height: state.props.height,
    tileHeight: state.props.tileHeight,
    sequences: state.sequences.raw,
    nrYTiles: state.sequenceStats.nrYTiles,
  }
}

export default msaConnect(
  mapStateToProps,
)(HTMLLabelsComponent);
// for testing
export {
  HTMLLabelsComponent as Labels,
}
