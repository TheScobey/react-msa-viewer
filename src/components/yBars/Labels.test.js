/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';

import dummySequences from '../../test/dummySequences';
import FakePositionStore from '../../test/FakePositionStore';
import { Labels } from './Labels';
import MSAViewer from '../MSAViewer';


it('renders properly', () => {
  const component = renderer.create(
    <FakePositionStore
      currentViewSequence={0}
    >
      <Labels
        nrYTiles={7}
        tileHeight={40}
        height={400}
        sequences={[...dummySequences]} />
    </FakePositionStore>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
