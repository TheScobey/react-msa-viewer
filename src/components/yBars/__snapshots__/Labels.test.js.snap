// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<div>
  <div
    style={
      Object {
        "height": 400,
        "overflow": "hidden",
        "position": "relative",
        "whiteSpace": "nowrap",
      }
    }
  >
    <div>
      <div
        style={
          Object {
            "height": 40,
          }
        }
      >
        sequence 1
      </div>
      <div
        style={
          Object {
            "height": 40,
          }
        }
      >
        sequence 2
      </div>
      <div
        style={
          Object {
            "height": 40,
          }
        }
      >
        sequence 3
      </div>
      <div
        style={
          Object {
            "height": 40,
          }
        }
      >
        sequence 4
      </div>
      <div
        style={
          Object {
            "height": 40,
          }
        }
      >
        sequence 5
      </div>
      <div
        style={
          Object {
            "height": 40,
          }
        }
      >
        sequence 6
      </div>
      <div
        style={
          Object {
            "height": 40,
          }
        }
      >
        sequence 7
      </div>
    </div>
  </div>
</div>
`;
